File: src/completed/java/euler0021.java
import java.util.ArrayList;

/**
 * Created by Holden on 12/06/2016. - COMPLETED
 *
 * Problem 21 - Amicable Numbers
 *
 */

public class Prob0021 {
    public static void main(String[] args) {
        long startTime = System.nanoTime();
        long answer = amicableNumbers(10000);
        long endTime = System.nanoTime();

        System.out.println(answer);

        long duration = (endTime - startTime);
        EulerTime.time(duration);
    }

    public static long amicableNumbers(int number) {
        int[] sumsArray = new int[number+1];

        for (int i = 2; i < number; i++) {
            ArrayList<Integer> properDivisors = EulerUtil.properDivisors(i);
            int sumDivisors = 0;
            for (int j: properDivisors) {
                sumDivisors += j;
            }

            sumsArray[i] = sumDivisors;
        }

        long sum = 0;

        for (int i = 2; i <= number; i++) {
            if ((sumsArray[i] < number) && (i == sumsArray[sumsArray[i]]) && (sumsArray[i] != i)) {
                sum += i;
            }
        }

        return sum;
    }
}

File: python/completed/euler0021.py
###
#
# Created by Holden on 12/18/2015
#
# COMPLETED on 10/17/2017
#
# Problem 21 - Amicable Numbers
#
###


def proper_divisors(n):
    divisors = []

    for i in range(1, n // 2 + 1):
        if n % i == 0:
            divisors.append(i)

    return divisors


def amicable_numbers(bound):
    sums_array = [-1, -1]

    for i in range(2, bound):
        divisors = proper_divisors(i)
        sum_divisors = 0
        for divisor in divisors:
            sum_divisors += divisor

        sums_array.append(sum_divisors)

    amicable_sum = 0

    for i in range(bound):
        if (sums_array[i] >= 0) and (sums_array[i] < bound) and (i == sums_array[sums_array[i]]) and (sums_array[i] != i):
            amicable_sum += sums_array[i]

    return amicable_sum

print(amicable_numbers(10000))
